mender-update/inventory.cpp

```diff
@@ -0,0 +1,171 @@
+}
+
+const string uri = "api/devices/v1/inventory/device/attributes";
+
+error::Error PushInventoryData(
+	const string &inventory_generators_dir,
+	const string &server_url,
+	http::Client &client,
+	events::EventLoop &loop,
+	APIResponseHandler api_handler) {
+	auto ex_inv_data = inv_parser::GetInventoryData(inventory_generators_dir);
+	if (!ex_inv_data) {
+		return ex_inv_data.error();
+	}
+
+	stringstream top_ss;
+	top_ss << "[";
+	auto inv_data = ex_inv_data.value();
+	for (const auto &kv : inv_data) {
+		top_ss << R"({"name":")";
+		top_ss << json::EscapeString(kv.first);
+		top_ss << R"(","value":)";
+		if (kv.second.size() == 1) {
+			top_ss << "\"" + json::EscapeString(kv.second[0]) + "\"";
+		} else {
+			stringstream items_ss;
+			items_ss << "[";
+			for (const auto &str : kv.second) {
+				items_ss << "\"" + json::EscapeString(str) + "\",";
+			}
+			auto items_str = items_ss.str();
+			// replace the trailing comma with the closing square bracket
+			items_str[items_str.size() - 1] = ']';
+			top_ss << items_str;
+		}
+		top_ss << R"(},)";
+	}
+	auto payload = top_ss.str();
+	// replace the trailing comma with the closing square bracket
+	payload[payload.size() - 1] = ']';
+
+	http::BodyGenerator payload_gen = [payload]() {
+		return make_shared<io::StringReader>(payload);
+	};
+
+	// TODO: APIRequest
+	auto req = make_shared<http::OutgoingRequest>();
+	req->SetAddress(path::Join(server_url, uri));
+	req->SetMethod(http::Method::PUT);
+	req->SetHeader("Content-Type", "application/json");
+	req->SetHeader("Content-Length", to_string(payload.size()));
+	req->SetHeader("Accept", "application/json");
+	req->SetBodyGenerator(payload_gen);
+
+	auto received_body = make_shared<vector<uint8_t>>();
+	return client.AsyncCall(
+		req,
+		[received_body, api_handler](http::ExpectedIncomingResponsePtr exp_resp) {
+			if (!exp_resp) {
+				log::Error("Request to push inventory data failed: " + exp_resp.error().message);
+				api_handler(exp_resp.error());
+			}
+
+			auto body_writer = make_shared<io::ByteWriter>(received_body);
+			auto resp = exp_resp.value();
+			auto content_length = resp->GetHeader("Content-Length");
+			auto ex_len = common::StringToLongLong(content_length.value());
+			if (!ex_len) {
+				log::Error("Failed to get content length from the inventory API response headers");
+				body_writer->SetUnlimited(true);
+			} else {
+				received_body->resize(ex_len.value());
+			}
+			resp->SetBodyWriter(body_writer);
+		},
+		[received_body, api_handler](http::ExpectedIncomingResponsePtr exp_resp) {
+			if (!exp_resp) {
+				log::Error("Request to push inventory data failed: " + exp_resp.error().message);
+				api_handler(exp_resp.error());
+			}
+
+			auto resp = exp_resp.value();
+			auto status = resp->GetStatusCode();
+			if (status == http::StatusOK) {
+				api_handler(error::NoError);
+			} else {
+				auto ex_err_msg = api::ErrorMsgFromErrorResponse(*(received_body.get()));
+				string err_str;
+				if (ex_err_msg) {
```

Comment:
Hmm, perhaps not a bad idea actually, doing it unconditionally :thinking:

Reply:
Do what unconditionally?

Reply:
You mean no matter what the status code is?

Reply:
What should the code do then? If the status code is 200:Success, but the response payload contains `"error": "Some error"`. File a bug? :wink:

Reply:
I was commenting on what you are doing here, which is doing it unconditionally, for all error values. Also undocumented ones.

Reply:
Oh! That's why it's best-effort -- it tries to extract the error and if it's not there, it uses the `StatusMessage` instead.

Reply:
Yes, exactly :)